app/models/user.rb

```diff
@@ -202,29 +204,27 @@ def send_devise_notification(notification, *args)
     devise_mailer.send(notification, self, *args).deliver_later
   end
 
-  def send_two_factor_authentication_code
-    # Would be used if we enable sending a code via SMS
-  end
-
-  def need_two_factor_authentication?(request)
+  def need_two_step_verification?
     otp_secret_key.present?
   end
 
-  def populate_otp_column
-    # Do nothing, this is only called on create and we don't
-    # necessarily want all users to have otp_secret_keys
-  end
-
   def authenticate_otp(code)
-    result = super(code)
+    totp = ROTP::TOTP.new(otp_secret_key)
+    result = totp.verify_with_drift(code, MAX_2SV_DRIFT_SECONDS)
+
     if result
       EventLog.record_event(self, EventLog::TWO_STEP_VERIFIED)
     else
       EventLog.record_event(self, EventLog::TWO_STEP_VERIFICATION_FAILED)
     end
+
     result
   end
 
+  def max_login_attempts?
```

Comment:
Is the naming of this method correct? It looks to be related specifically to 2SV, so should probably be named to indicate that.